### PR TITLE
refactor(yaml): make `readDocument()` and `readDocuments()` generator functions

### PR DESCRIPTION
**Pull request by timreichen**
- makes `readDocuments()` and add `readDocuments()` generator functions

This will prevent parsing all documents in `load()` before throwing an error.

**Note**: The performance can be enhanced in a future PR by checking whether the state.position is at the end of the state.input instead of parsing a second document. There are some unknown state changes going on somewhere during the parsing. That is why this approach doesn't work atm.

### CODE REVIEW

File: yaml/_loader/loader.ts [hunk 1]
```diff
@@ -1565,7 +1565,7 @@ function composeNode(
   return state.tag !== null || state.anchor !== null || hasContent;
 }
 
-function readDocument(state: LoaderState) {
+function* readDocument(state: LoaderState) {
```
[the remark this answers is not in the document]
Comment (timreichen) [reply yaml/_loader/loader.ts]: I agree, done.

File: yaml/_loader/loader.ts [hunk 2]
```diff
@@ -1765,31 +1773,23 @@ export function loadDocuments(
...
-
-  while (state.position < state.length - 1) {
-    readDocument(state);
+  const documents = [];
+  for (const document of readDocuments(state)) {
+    documents.push(document);
   }
-
-  return state.documents;
+  return documents;
```
Comment (iuioiua) [inline yaml/_loader/loader.ts]:
```suggestion
  return [...readDocuments(state)];
```

**Review by iuioiua — APPROVED** [review]

**Comment by kt3k** [comment]
Should we also remove `documents` field of `LoaderState`?

**Review by kt3k — APPROVED** [review]
LGTM

**Pull request merged**